### ISSUE TITLE
Implement DBField validation

### ISSUE THREAD

**Issue by emteknetnz** [silverstripe/silverstripe-framework#11403]
[Consolidate validation systems](https://github.com/silverstripe/silverstripe-framework/issues/11391) contained a [POC](https://github.com/silverstripe/silverstripe-framework/pull/11402) that added DBField validation via a new shared FIeldsValidator class. We agreed that we should proceed with implementing this for CMS 6

## Acceptance criteria
- [x] Technical approach taken in the [POC](https://github.com/silverstripe/silverstripe-framework/pull/11402) is used
- [x] Change the double array config from the POC to instead allow options A), B) and C)
- [x] Types of validation that could be handled on FieldValidator's are identified, e.g. Email, URL, Int size, etc - note the existing [data types](https://docs.silverstripe.org/en/5/developer_guides/model/data_types_and_casting/)
- [ ] ~Existing FormField validate methods are updated to use FieldValidator's where possible~ Split off as [new issue](https://github.com/silverstripe/silverstripe-framework/issues/11422)
- [x] Some new DB fields are created with validation e.g. DBEmail, DBUrl, DBIp
- [x] New DB fields are autoscaffold existing equivalent FormField if one exists e.g. EmailField, though DBIp would just be a TextField
- [x] New DBFields are listed in docs on the [Data types](https://docs.silverstripe.org/en/5/developer_guides/model/data_types_and_casting/) page
- [x] DBField validation is mentioned in the developer docs
- [x] 6.0.0 changelog calls out that data will no longer be inserted and truncated if it exceeds varchar character lengths, etc, instead it will throw a validation exception
- [x] Discuss / make follow up cards for "things that we might want to consider changing" on the framework PR

## New issues created
- https://github.com/silverstripe/silverstripe-framework/issues/11422
- https://github.com/silverstripe/silverstripe-framework/issues/11423
- https://github.com/silverstripe/silverstripe-framework/issues/11445

## Kitchen sink CI CMS 6
- https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/11693442025

## CMS 5 PRs
- https://github.com/silverstripe/silverstripe-assets/pull/650
- https://github.com/silverstripe/silverstripe-framework/pull/11427
- https://github.com/silverstripe/developer-docs/pull/603

## CMS 6 PRs
- https://github.com/silverstripe/silverstripe-assets/pull/651
- https://github.com/silverstripe/silverstripe-framework/pull/11408
- https://github.com/silverstripe/silverstripe-cms/pull/3013
- https://github.com/silverstripe/silverstripe-assets/pull/648
- https://github.com/silverstripe/silverstripe-contentreview/pull/263
- https://github.com/silverstripe/silverstripe-frameworktest/pull/206
- https://github.com/symbiote/silverstripe-advancedworkflow/pull/558
- https://github.com/silverstripe/developer-docs/pull/596
- https://github.com/silverstripe/silverstripe-userforms/pull/1342

**Comment by emteknetnz** [comment]
## FieldValidators

---

### BooleanValidator
- Validates that a value is either 1 or 0 - or true/false (gets converted before DB insert?)
- Note that DBBoolean datatype is tinyint

#### DBFields
- DBBoolean (extends FormField)

#### FormFields
- CheckboxField

---

### CompositeValidator
- Calls validate() on child fields

#### DBFields
- DBMoney (extends DBComposite)
- DBPolymorphicForeignKey (extends DBComposite)

#### FormFields
- MoneyField
---

### EnumValidator
- Validate the value is in an fixed set of value

#### DBFields
- DBEnum (extends FormField)
- DBMultiEnum (extends Enum)
- DBClassName  (extends Enum)
- DBClassNameVarchar

#### FormFields
- None

---

### DecimalValidator
- Validates a value is between minValue and maxValue with respect to the maximum size of the database column
- Will need to be able to handle signed values (i.e. less than 0)
- Should not validate number of decimal places, allow database to handle rounding to column precision

#### DBFields
- DBDecimal (extends FormField)
- DBPercentage (extends Decimal)
- DBCurrency (extends Decimal)
- ~~DBFloat (extends FormField)~~ too complicated
- ~~DBDouble (extends FormField)~~ too complicated

#### FormFields
- None

---

### DateValidator
- Validate a date is formatted using DBDate::ISO_DATE = 'y-MM-dd';

#### DBFields
- DBDate (extends FormField)

#### FormFields
- DateField

---

### DateTimeValidator
- Validate a datetime is formatted using public const ISO_DATETIME = 'y-MM-dd HH:mm:ss';

#### DBFields
- DBDateTime (extends DBDate)

#### FormFields
- DateTimeField

---

### EmailValidator
- Validates the string is a valid email

#### DBFields
- DBEmail (extends DBVarchar) (new field)

#### FormFields
- EmailField

---

### IntValidator
- Validates a value is between minValue and maxValue with respect to the maximum size of the database field
- Will need to be able to handle signed values (i.e. less than 0)
- Validate there is no decimal value

#### DBFields
- DBInt (extends FormField)
- DBBigInt (extends DBInt)
- DBYear (extends FormField)
- DBForiegnKey (extends DBInt)
- DBPrimaryKey (extends DBInt)

#### FormFields
- None

---

### IpValidator
- Validates the string is a valid IP address

#### DBFields
- DBIp (extends DBVarchar) (new field)

#### FormFields
- None

---

### LocaleValidator (extends StringValidator)
- Validate the string matches a standard locale pattern

#### DBFields
- DBLocale (extends DBVarchar)

#### FormFields
- None

---

### StringValidator
- Validate the length of the string is between minLength and maxLength

#### DBFields
- DBVarchar (extends DBString)
- DBText (extends DBString)
- DBVarchar (extends DBString)
- DBText (extends DBString)
- DBClassName (extends DBVarchar)
- DBHTMLFragment (class is DBHTMLText without shortcode parsing)
- DBHTMLText (extends DBText) (with shortcode parsing)
- DBHTMLVarchar (extends DBVarchar)

#### FormFields
- TextField
- TextareaField

---

### TimeValidator
- Validate a time is formatted using DBTime::ISO_TIME = 'HH:mm:ss';

#### DBFields
- DBTime (extends FormField)

#### FormFields
- TimeField

---

### UrlValidator
- Validates the string is a valid url

#### DBFields
- DBUrl (extends DBVarchar) (new field)

#### FormFields
- UrlField

**Comment by GuySartorelli** [comment]
@emteknetnz Don't forget to do a CMS 5 changelog PR for the deprecations.

**Comment by emteknetnz** [comment]
Done

**Comment by GuySartorelli** [comment]
@emteknetnz Most of the PRs are still in draft? Does the assets one have to be merged before the others can be properly reviewed and merged? If so please add a note to that effect to the issue so it's clear what's going on.

**Comment by emteknetnz** [comment]
Yes sorry, merge that one first, I'll need to rebase the other assets PR once it's merge. I'll mark all of the draft PRs to ready for review after that's done

**Comment by GuySartorelli** [comment]
@emteknetnz I can't test these PRs because they need to be rebased after the recent template work got merged in.
Can you please rebase the PRs?

**Comment by emteknetnz** [comment]
Has been done

**Comment by GuySartorelli** [comment]
PRs merged
